### PR TITLE
Changes for the second session

### PR DESCRIPTION
**Pull request by KipHamiltons**
I added a bunch of things, fixed a bunch of typos, etc.

There are a couple of examples which need finishing too. Do these at your own leisure - I am not 100% on where you were going with them **(note that these line references are on THIS branch. I don't know the line numbers on `master`)**

- [ ] line 748 example has `if` commented out
- [ ] line 863 is incomplete
- [ ] line 677 example is strange to me. It doesn't illustrate anything about l-values or r-values to me.

### CODE REVIEW

File: cpp/moderncpp.Rmd [hunk 1]
```diff
@@ -221,27 +232,34 @@ In this example we cannot convert from `simple_container<int>` to `simple_contai
...
 
-There are some caveats with raw pointers, but we don't use raw pointers, so if that ever is the case there are many other resources explaining what is happening.
+There are some caveats with raw pointers (see definition in [ownership](#raw-pointers)), but we _very rarely_ use raw pointers.
+If you're confused about such a case, there are many other resources explaining what is happening.
+
+## How to Read `const` Specifiers
+
+The rule when reading types in C++ is that you should read "from right to left".
+This manifests as `const` applying to the thing immediately to its left.
+If there isn't anything to the left, it applies to the thing on its right - please don't use ["East const"](https://isocpp.org/wiki/faq/const-correctness#const-ref-alt) though...
```
Comment (CMurtagh-LGTM) [inline cpp/moderncpp.Rmd]:
```suggestion
e.g. `cont int*` and `int const*` are both a pointer to a constant int, `int* const` is a const pointer to an int and `const int* const` or `int const* const` is a const pointer to a const int.
```

File: cpp/moderncpp.Rmd [hunk 2]
```diff
@@ -809,28 +868,38 @@ int main(){
...
+A `std::deque` implements a deque (double-ended queue).
+The expansion of capacity and insertion or removal of elements at the ends is much faster than a vector, but takes up more memory, and accesses are slower.
+Usually, a vector is better as we often do more reads than insertion and removals.
 
 ## Map
 
-This is an associative data structure, it associates values to keys. It can either be a red-black tree, [std::map](https://en.cppreference.com/w/cpp/container/map), or a hash map, [std::unordered_map](https://en.cppreference.com/w/cpp/container/unordered_map).
+This is an associative data structure which associates values to keys.
+It can either be a red-black tree ([std::map](https://en.cppreference.com/w/cpp/container/map)), or a hash map ([std::unordered_map](https://en.cppreference.com/w/cpp/container/unordered_map)).
+Maps are bad for cache-locality and they should be avoided where possible in performance critical code.
```
Comment (CMurtagh-LGTM) [inline cpp/moderncpp.Rmd:904]: Just say they are slow, no one will know about cache locality.

**Review by CMurtagh-LGTM — CHANGES_REQUESTED** [review]

File: cpp/moderncpp.Rmd (hunk 1 above)
Comment (KipHamiltons) [reply cpp/moderncpp.Rmd]: Resuggesting your suggestion with a typo fix, so I can commit it.
```suggestion
e.g. `const int*` and `int const*` are both a pointer to a constant int, `int* const` is a const pointer to an int and `const int* const` or `int const* const` is a const pointer to a const int.
```

File: cpp/moderncpp.Rmd (hunk 2 above)
Comment (KipHamiltons) [reply cpp/moderncpp.Rmd:904]: I think that's implied by the rest of the sentence. I think this detail is worth keeping.

**Pull request merged**